### ISSUE TITLE
[Bug]: AC0031 and AC0032 are contradictory for table permissions 

### ISSUE THREAD

**Issue by ralf-escher** [ALCops/Analyzers#295]
### Analyzer

ApplicationCop
**AL version:** BC 28
**ALCop version:** 0.8.3

### Issue type

False positive (diagnostic fires when it shouldn't)

### Rule ID

AC0031 and AC0032

### Description

AC0031 and AC0032 are contradictory for table permissions that are required at runtime but not directly visible to the static analyzer. Declaring the permission triggers AC0032; removing it triggers AC0031. There is no way to satisfy both rules simultaneously without a #pragma suppression.

**To reproduce**

Consider a codeunit that receives a Record "Posted Whse. Shipment Line" as a parameter and reads a field from it:

With the permission declared: AC0032 fires — "Permission 'tabledata Posted Whse. Shipment Line' is declared but this object performs no database operations on table 'Posted Whse. Shipment Line'."
With the permission removed: AC0031 fires — "Permission for table 'Posted Whse. Shipment Line' is missing."
Expected behavior

Passing a record as a parameter and accessing its fields constitutes a read operation. AC0031/AC0032 should recognise record parameters as an implicit = R usage and not flag the permission as unused.

**Actual behavior**

Both AC0031 and AC0032 fire depending on whether the permission is declared or not. The only escape is:

```
#pragma warning disable AC0032
tabledata "Posted Whse. Shipment Line" = R,   // needed at runtime via record parameter
#pragma warning restore AC0032
```

**Additional context**

This also occurs for permissions required through:

Event subscriber parameters carrying a record variable
Indirect calls where the record is passed through a chain of local procedures

### AL code to reproduce

```codeunit 50000 "My Codeunit"
{
    Permissions =
        tabledata "Posted Whse. Shipment Line" = R;  // AC0032 fires here

    procedure DoSomething(ShipmentLine: Record "Posted Whse. Shipment Line")
    begin
        // Reading a field value — no explicit Get/FindSet/FindFirst call
        if ShipmentLine."No." <> '' then
            DoWork(ShipmentLine."No.");
    end;
}
```

**Comment by Arthurvdv** [comment]
I could use some help to reproduce this issue. If I create a small code example the AC0031 diagnostic isn't raised. What am I missing here?

```AL
codeunit 50100 MyCodeunit
{
    procedure MyProcedure(ShipmentLine: Record "Posted Whse. Shipment Line")
    begin
        if ShipmentLine."No." <> '' then
            this.DoWork(ShipmentLine."No.");
    end;

    local procedure DoWork(RefNo: Code[20])
    begin
    end;
}
```

**Comment by ralf-escher** [comment]
Hi @Arthurvdv,
I checked two projects and was able to remove the permissions that triggered the AC0032 warnings—thanks for the hint!
The remaining issue concerns reports. I could try to build a minimal example to illustrate the problem if that would help.
Let me know if you need further details.

**Comment by Arthurvdv** [comment]
@ralf-escher With the update of `v0.8.4` of ALCops, maybe also your issues are resolved?

Would you be able to update to this version and let me know if it fixes both issues on your side as well.

**Comment by tkroes** [comment]
@Arthurvdv I am using v0.9.0-beta.1 and found a way to reproduce this issue. It seems to do specifically (in this case) with read functions that don't hard require parentheses. The following snippet reproduces it for me:
```al
codeunit 50001 PermissionTest
{
    Permissions = tabledata "Sales Invoice Line" = R;

    trigger OnRun()
    var
        SalesInvoiceLine: Record "Sales Invoice Line";
    begin
        Message('%1', SalesInvoiceLine.Count); // Note no Count(), which works correctly
    end;
}
```
This sample throws AC0032 on the permission declaration and AC0031 on the `.Count` call if I remove it.

**Comment by Arthurvdv** [comment]
@tkroes Thanks for reporting this and for providing the `.Count` example.

I've updated the analyzer to resolve the contradiction between AC0031 and AC0032 for this scenario. From a Roslyn perspective, `.Count` and `.Count()` are two different constructs, while the AL compiler currently accepts both forms. This makes the analysis a bit more challenging, especially if AL ever introduces actual properties on objects in the future.

Would you mind giving this a try with the ALCops beta channel? The fix is available in `0.9.0-beta.2`. I'd be interested to hear whether it also resolves the issue in your environment.

**Comment by tkroes** [comment]
@Arthurvdv I can indeed confirm that the issue is resolved in version 0.9.0-beta.2. Thank you for the quick update!

**Comment by Arthurvdv** [comment]
I believe I've resolved the issues here with the release of version `v0.9.0` of ALCops.

I'm going ahead and close the issue here, but feel free to reopen if it's not working as expected.

**Comment by CGJenDev** [comment]
I am on `v0.9.0` and it is still doing this for random tables, however it seems to only do this with the `M` permission tag for a table. @Arthurvdv 

Please let me know if I need to open a new issue, the context and code examples would be the same as this issue.

**Comment by Arthurvdv** [comment]
@CGJenDev I've made a few small changes to this rule in #339 and #353, could you try ALOps `1.0.0-beta.1` to see if these changes resolve the issue?

If the problem persists, it would indeed be best to open a new issue where I'd be happy to look into it.

**Comment by CGJenDev** [comment]
@Arthurvdv All working, thank you :)
